src/Adapter/Product/CommandHandler/UpdateProductSeoHandler.php

```diff
@@ -0,0 +1,145 @@
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\RedirectOption;
+use Product;
+
+/**
+ * Handles @var UpdateProductSeoCommand using legacy object model
+ */
+class UpdateProductSeoHandler extends AbstractProductHandler implements UpdateProductSeoHandlerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(UpdateProductSeoCommand $command): void
+    {
+        $product = $this->getProduct($command->getProductId());
+        $this->fillUpdatableFieldsWithCommandData($product, $command);
+
+        $this->performUpdate($product, CannotUpdateProductException::FAILED_UPDATE_SEO);
+    }
+
+    /**
+     * @param Product $product
+     * @param UpdateProductSeoCommand $command
+     */
+    private function fillUpdatableFieldsWithCommandData(Product $product, UpdateProductSeoCommand $command): void
+    {
+        $redirectOption = $command->getRedirectOption();
+
+        if (null !== $redirectOption) {
+            $this->fillRedirectOptionValues($product, $redirectOption);
+        }
+
+        if (null !== $command->getLocalizedMetaDescriptions()) {
+            $product->meta_description = $command->getLocalizedMetaDescriptions();
+            $this->validateLocalizedField($product, 'meta_description', ProductConstraintException::INVALID_META_DESCRIPTION);
+            $this->fieldsToUpdate['meta_description'] = true;
+        }
+
+        if (null !== $command->getLocalizedMetaTitles()) {
+            $product->meta_title = $command->getLocalizedMetaTitles();
+            $this->validateLocalizedField($product, 'meta_title', ProductConstraintException::INVALID_META_TITLE);
+            $this->fieldsToUpdate['meta_title'] = true;
+        }
+
+        if (null !== $command->getLocalizedLinkRewrites()) {
+            $product->link_rewrite = $command->getLocalizedLinkRewrites();
+            $this->validateLocalizedField($product, 'link_rewrite', ProductConstraintException::INVALID_LINK_REWRITE);
+            $this->fieldsToUpdate['link_rewrite'] = true;
+        }
+    }
+
+    /**
+     * @param Product $product
+     * @param RedirectOption $redirectOption
+     *
+     * @throws CategoryNotFoundException
+     * @throws ProductException
+     * @throws ProductNotFoundException
+     */
+    private function fillRedirectOptionValues(Product $product, RedirectOption $redirectOption): void
+    {
+        $redirectType = $redirectOption->getRedirectType();
+        $redirectTarget = $redirectOption->getRedirectTarget();
+
+        if ($redirectType->isProductType()) {
+            $this->assertProductExists($redirectTarget->getValue());
+        } elseif (!$redirectType->isTypeNotFound() && !$redirectTarget->isNoTarget()) {
```

Comment:
I wonder if it would be useful to have `isCategoryType()` maybe worth adding it before feature freeze since after the class will be "locked" (suggestion for future PR)

Reply:
As it was said multiple times, we shouldn't implement "Future methods" expecting that they might be useful for someone some day, so I guess not :man_shrugging:

Reply:
I've changed my mind, yes you are right. We should improve it in some other pr, it will become much cleaner to use `$redirectType->isCategoryType()` instead of `$redirectType->isTypeNotFound()`